### PR TITLE
Draft: RISC-V support

### PR DESCRIPTION
**Pull request by TobiasFaller**
This will be my attempt to implement RISC-V support (#735) for Pwndbg.
Capstone has RISC-V support only in the `next` branch, so this pull request will stay open until the new release version 5.0.0 is released.
Feel free to give feedback.

### CODE REVIEW

**Comment by TobiasFaller** [comment]
There are currently some bugs in Capstone / the Python bindings.
This should be fixed soon.

See:
https://github.com/aquynh/capstone/pull/1690
https://github.com/aquynh/capstone/pull/1691

**Comment by TobiasFaller** [comment]
This post has been moved to a separate pull request https://github.com/pwndbg/pwndbg/pull/830.

File: pwndbg/regs.py [hunk 1]
```diff
@@ -240,6 +240,46 @@ def __iter__(self):
...
+# f10–11    => fa0–1  (FP arguments/return values)
+# f12–17    => fa2–7  (FP arguments)
+# f18–27    => fs2–11 (FP saved registers)
+# f28–31    => ft8–11 (FP temporaries)
+riscv = RegisterSet( pc      = 'pc',
+                     frame   = 'fp',
+                     stack   = 'sp',
+                     retaddr = 'ra',
+                     gpr     = ('ra', 'gp', 'tp', )
+                               + ('t0', 't1', 't2', 'fp', 's1', )
```
Comment (disconnect3d) [inline pwndbg/regs.py]: Maybe:
```python
                     gpr     = ('ra', 'gp', 'tp', \
                                't0', 't1', 't2', 'fp', 's1', \
```
o even without the `\` character.

The `tuple(...)` should also probably be pre-computed & hardcoded (same for other archs, heh!)

Comment (TobiasFaller) [reply pwndbg/regs.py]: Ok, I will open a separate pull-request for the other architectures.

**Comment by disconnect3d** [comment]
@TobiasFaller there seem to be a conflict; can u fix it soo we can merge it? :)

**Comment by TobiasFaller** [comment]
> @TobiasFaller there seem to be a conflict; can u fix it soo we can merge it? :)

Yeah, sure. As this is still a WIP I have not rebased it yet.
I will update this MR after Feb. the 18th.

**Comment by TobiasFaller** [comment]
Updates: aquynh/capstone#1690 has now been merged too.

**Comment by TobiasFaller** [comment]
Update: There has not been a release of aquynh/capstone with RISC-V support yet. Maybe it is time to move to a more active project / fork?

**Comment by disconnect3d** [comment]
> Maybe it is time to move to a more active project / fork?

Is there actually a fork that is more active or/and supports RISC-V?

**Comment by lyctw** [comment]
Hi
Does RISC-V support still wait for Capstone 5.0.0 release?

**Comment by TobiasFaller** [comment]
> Hi Dose RISC-V support still wait for Capstone 5.0.0 release?

Yes, but the first release candidates seem to have been released ten days ago which means it's probably not long before a stable release will be done.

Please keep in mind that this RISC-V extension's code has not been fully tested yet and can only be considered a prototype.
I'll probably get back to this implementation in the next weeks.

**Comment by disconnect3d** [comment]
@TobiasFaller ping :)

**Comment by TobiasFaller** [comment]
> @TobiasFaller ping :)

Still waiting for a stable capstone 5.0 release.

**Comment by lonnywong** [comment]
@disconnect3d @TobiasFaller 

It's been a long time. Can we support RISC-V without capstone?

How about make a simple disassembler by calling
```py
instr = gdb.execute("x/i %#x" % address, to_string=True)
```

**Comment by FantasqueX** [comment]
Really want this feature. Encounter some difficulty when debugging on riscv64 hardware without pwndbg. Thanks everyone working on this issue in advance.

Update: I'll try the one in master branch.

**Comment by disconnect3d** [comment]
@FantasqueX pls let me know if and how the dev/latest version works for u!

**Comment by disconnect3d** [comment]
@TobiasFaller fwiw we could potentially use the release candidate version too. I'm rather okay with it, as long as we give a choice to users to switch between the current solution and capstone based one.

**Comment by TobiasFaller** [comment]
> @TobiasFaller fwiw we could potentially use the release candidate version too. I'm rather okay with it, as long as we give a choice to users to switch between the current solution and capstone based one.

@disconnect3d Okay, I'll see what I can do and how to make the release candidate version usable. I'll need to validate and rebase the current implementation though, so it will take time.

**Comment by Xeonacid** [comment]
> @disconnect3d Okay, I'll see what I can do and how to make the release candidate version usable. I'll need to validate and rebase the current implementation though, so it will take time.

Hi. How is it going? Is there anything I can do to help? I'm very interested in this work.

**Comment by disconnect3d** [comment]
@Xeonacid Hi! Sure, you can help :)

We added some basic support for riscv:rv64 in #1588.

To work with RISC-V further we need to:
1. Add other RISC-V architecture variants: mainly, make sure we map registers properly and that we have a corresponding Capstone disassebler. If there is no Capstone disassembler, we have to fallback to the GDB one somehow.
2. Ideally, add some basic qemu tests for those ("does it launch without crash", "does it provide correct registers, disassembly and stack output", etc)
3. Extend the support to enhance the disassembly: this is what `DisassemblyAssistant` can be used for. Note that this feature overall may need some overwork as it was mainly designed for x86/x64 stuff.

Generally, I think we should close this draft since it is conflicted and not worked upon anyway, but one can base upon this and extend Pwndbg further.

**Comment by Xeonacid** [comment]
> @Xeonacid Hi! Sure, you can help :)
> 
> We added some basic support for riscv:rv64 in #1588.
> 
> To work with RISC-V further we need to:
> 
> 1. Add other RISC-V architecture variants: mainly, make sure we map registers properly and that we have a corresponding Capstone disassebler. If there is no Capstone disassembler, we have to fallback to the GDB one somehow.
> 2. Ideally, add some basic qemu tests for those ("does it launch without crash", "does it provide correct registers, disassembly and stack output", etc)
> 3. Extend the support to enhance the disassembly: this is what `DisassemblyAssistant` can be used for. Note that this feature overall may need some overwork as it was mainly designed for x86/x64 stuff.
> 
> Generally, I think we should close this draft since it is conflicted and not worked upon anyway, but one can base upon this and extend Pwndbg further.

Thanks for the information! I will work on this.

**Comment by disconnect3d** [comment]
> Thanks for the information! I will work on this.

Great! If you have any questions, you can also ask on our Discord: https://discord.gg/9rrbG3yt

**Comment by disconnect3d** [comment]
I am closing this PR for now

**Pull request closed without merging**